File: Math/CPP/number_of_days_between_two_dates.cpp
#include <algorithm>
#include <string>
using namespace std;

class Solution {
public:
    int daysBetweenDates(string date1, string date2) {
        if (date1 > date2) swap(date1, date2); // Ensure date1 <= date2

        int date1Y = stoi(date1.substr(0, 4)),
            date1M = stoi(date1.substr(5, 2)),
            date1D = stoi(date1.substr(8, 2));
        int date2Y = stoi(date2.substr(0, 4)),
            date2M = stoi(date2.substr(5, 2)),
            date2D = stoi(date2.substr(8, 2));

        return daysFromStart(date2Y, date2M, date2D) - daysFromStart(date1Y, date1M, date1D);
    }

private:
    int daysFromStart(int year, int month, int day) {
        int days[13] = {0, 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31};
        int totalDays = 0;

        // Add days for full years
        for (int y = 1971; y < year; ++y) {
            totalDays += 365 + (isLeap(y) ? 1 : 0);
        }

        // Add days for full months of the current year
        for (int m = 1; m < month; ++m) {
            totalDays += days[m];
            if (m == 2 && isLeap(year)) totalDays += 1; // Add leap day
        }

        // Add remaining days of the current month
        totalDays += day;

        return totalDays;
    }

    bool isLeap(int year) {
        return (year % 4 == 0) && (year % 100 != 0 || year % 400 == 0);
    }
};

File: Math/Python/number_of_days_between_two_dates.py
class Solution:
    def isLeap(self, year) -> bool:
        return year % 4 == 0 and (year % 100 != 0 or year % 400 == 0)
    def daysFromStart(self, year, month, day) -> int:
        days = [0,31,28,31,30,31,30,31,31,30,31,30,31]
        totalDays = 0
        for y in range(1971, year): totalDays += 366 if self.isLeap(y) else 365
        for m in range(1, month): totalDays += days[m] + 1 if self.isLeap(year) and m == 2 else days[m]
        totalDays += day
        return totalDays
    def daysBetweenDates(self, date1: str, date2: str) -> int:
        if date1 > date2: date1, date2 = date2, date1
        y1, m1, d1 = map(int, date1.split("-"))
        y2, m2, d2 = map(int, date2.split("-"))
        return self.daysFromStart(y2, m2, d2) - self.daysFromStart(y1, m1, d1)
